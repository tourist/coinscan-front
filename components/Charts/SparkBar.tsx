import { useEffect } from 'react';
import Box from '@mui/material/Box';
import {
  extent as d3_extent,
  select as d3_select,
  scaleLinear as d3_scaleLinear,
  scaleBand as d3_scaleBand,
} from 'd3';
import type { DataPoint } from '../../utils/charts';

function renderSparkBar(
  selector: string,
  data: DataPoint<bigint | number>[],
  colors: [positiveColor: string, negativeColor: string] = ['green', 'tomato']
) {
  const DATA_EXTENT = d3_extent(data, (d) => Number(d.count));

  if (DATA_EXTENT[0] === undefined || DATA_EXTENT[1] === undefined) {
    return;
  }

  const WIDTH = 200;
  const HEIGHT = 50;
  const DATA_MAX = DATA_EXTENT[1];
  const DATA_MIN = DATA_EXTENT[0];
  const SIDE_MAX = Math.max(Math.abs(DATA_MIN), Math.abs(DATA_MAX));

  const x = d3_scaleBand()
    .domain(data.map((d) => d.id))
    .range([0, WIDTH])
    .paddingInner(0.5)
    .paddingOuter(0.5);
  const y = d3_scaleLinear().domain([-SIDE_MAX, SIDE_MAX]).range([HEIGHT, 0]);

  // clear existing renders
  d3_select(selector).selectAll('svg').remove();

  const svg = d3_select(selector)
    .append('svg')
    .attr('viewBox', '0 0 200 50')
    .attr('preserveAspectRatio', 'xMidYMid');

  svg
    .append('line')
    .attr('x1', 0)
    .attr('y1', y(0))
    .attr('x2', WIDTH)
    .attr('y2', y(0))
    .attr('stroke', 'white')
    .attr('stroke-opacity', 0.1)
    .attr('shape-rendering', 'crispEdges');

  svg.append('g');
  svg
    .selectAll('.bar')
    .data(data.filter((i) => i.count != BigInt(0)))
    .enter()
    .append('rect')
    .attr('class', 'bar')
    .attr('x', (d) => x(d.id) || null)
    .attr('y', (d) => (d.count > 0 ? y(Number(d.count)) - 1 : y(0) + 1))
    .attr('width', x.bandwidth)
    .attr('height', (d) => Math.abs(y(Number(d.count)) - y(0)))
    .attr('fill', (d) => (d.count > 0 ? colors[0] : colors[1]));
}

type SparkBarProps = {
  id: number | string;
  data: DataPoint<bigint | number>[];
};

const SparkBar = ({ id, data }: SparkBarProps) => {
  useEffect(() => renderSparkBar(`#spark-bar${id}`, data), [id, data]);
  if (data.length === 0) return null;
  return (
    <Box
      sx={{
        width: 150,
        display: 'flex',
        justifyContent: 'center',
        alignItems: 'center',
        svg: {
          width: '100%',
          height: '40px',
        },
      }}
      id={`spark-bar${id}`}
    ></Box>
  );
};

export default SparkBar;
