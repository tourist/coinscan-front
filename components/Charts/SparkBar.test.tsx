import { render, screen } from '@testing-library/react';

import SparkBar from './SparkBar';

test('renders with single value', async () => {
  const { container } = render(
    <SparkBar
      id="color-scale-test"
      data={[
        { id: '1660262400', count: 0 },
        { id: '1660348800', count: 0 },
        { id: '1660435200', count: 0 },
        { id: '1660521600', count: 0 },
        { id: '1660608000', count: 0 },
        { id: '1660694400', count: 0 },
        { id: '1660780800', count: 0 },
        { id: '1660867200', count: 2000000000000 },
        { id: '1660953600', count: 0 },
        { id: '1661040000', count: 0 },
        { id: '1661126400', count: 3000000000000 },
        { id: '1661212800', count: 0 },
        { id: '1661299200', count: 0 },
        { id: '1661385600', count: -9000000000000 },
      ]}
    />
  );
  expect(document.querySelectorAll('rect').length).toEqual(3);
  expect(container).toMatchSnapshot();
});

test('not render when empty data array is passed', async () => {
  const { container } = render(<SparkBar id="color-scale-test" data={[]} />);
  expect(container).toBeEmptyDOMElement();
});
