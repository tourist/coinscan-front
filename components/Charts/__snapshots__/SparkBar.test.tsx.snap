// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders with single value 1`] = `
<div>
  <div
    class="MuiBox-root css-1pwmyt7"
    id="spark-barcolor-scale-test"
  >
    <svg
      preserveAspectRatio="xMidYMid"
      viewBox="0 0 200 50"
    >
      <line
        shape-rendering="crispEdges"
        stroke="white"
        stroke-opacity="0.1"
        x1="0"
        x2="200"
        y1="25"
        y2="25"
      />
      <g />
      <rect
        class="bar"
        fill="green"
        height="5.555555555555557"
        width="6.896551724137931"
        x="103.44827586206897"
        y="18.444444444444443"
      />
      <rect
        class="bar"
        fill="green"
        height="8.333333333333332"
        width="6.896551724137931"
        x="144.82758620689654"
        y="15.666666666666668"
      />
      <rect
        class="bar"
        fill="tomato"
        height="25"
        width="6.896551724137931"
        x="186.20689655172413"
        y="26"
      />
    </svg>
  </div>
</div>
`;
